service/controller/resource/azureclusterconfig/create.go

```diff
@@ -0,0 +1,274 @@
+	if err != nil {
+		return corev1alpha1.AzureClusterConfig{}, microerror.Mask(err)
+	}
+
+	failureDomains, err := getAvailabilityZones(masters)
+	if err != nil {
+		return corev1alpha1.AzureClusterConfig{}, microerror.Mask(err)
+	}
+
+	clusterConfig := corev1alpha1.AzureClusterConfig{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "AzureClusterConfig",
+			APIVersion: "core.giantswarm.io/v1alpha1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      clusterConfigName(key.ClusterID(&cluster)),
+			Namespace: metav1.NamespaceDefault,
+			Labels: map[string]string{
+				label.ClusterOperatorVersion: clusterOperatorVersion,
+			},
+		},
+		Spec: corev1alpha1.AzureClusterConfigSpec{
+			Guest: corev1alpha1.AzureClusterConfigSpecGuest{
+				ClusterGuestConfig: corev1alpha1.ClusterGuestConfig{
+					AvailabilityZones: len(failureDomains),
+					DNSZone:           dnsZoneFromAPIEndpoint(azureCluster.Spec.ControlPlaneEndpoint.Host),
+					ID:                key.ClusterID(&cluster),
+					Name:              key.ClusterID(&cluster),
+					Owner:             key.OrganizationID(&cluster),
+					ReleaseVersion:    key.ReleaseVersion(&cluster),
+					VersionBundles:    componentsToClusterGuestConfigVersionBundles(cc.Release.Release.Spec.Components),
+				},
+				CredentialSecret: corev1alpha1.AzureClusterConfigSpecGuestCredentialSecret{
+					Name:      credentialSecret.Name,
+					Namespace: credentialSecret.Namespace,
+				},
+				Masters: nil,
+				Workers: nil,
+			},
+			VersionBundle: corev1alpha1.AzureClusterConfigSpecVersionBundle{
+				Version: clusterOperatorVersion,
+			},
+		},
+	}
+
+	for _, master := range masters {
+		m := corev1alpha1.AzureClusterConfigSpecGuestMaster{
+			AzureClusterConfigSpecGuestNode: corev1alpha1.AzureClusterConfigSpecGuestNode{
+				ID:     master.Name,
+				VMSize: master.Spec.VMSize,
+			},
+		}
+
+		clusterConfig.Spec.Guest.Masters = append(clusterConfig.Spec.Guest.Masters, m)
+	}
+
+	for _, worker := range workers {
+		w := corev1alpha1.AzureClusterConfigSpecGuestWorker{
+			AzureClusterConfigSpecGuestNode: corev1alpha1.AzureClusterConfigSpecGuestNode{
+				ID:     worker.Name,
+				VMSize: worker.Spec.VMSize,
+			},
+			Labels: worker.Labels,
+		}
+
+		clusterConfig.Spec.Guest.Workers = append(clusterConfig.Spec.Guest.Workers, w)
+	}
+
+	return clusterConfig, nil
+}
+
+func clusterConfigName(clusterID string) string {
+	return fmt.Sprintf("%s-%s", clusterID, "azure-cluster-config")
+}
+
+func dnsZoneFromAPIEndpoint(apiEndpointHost string) string {
+	domainLabels := strings.Split(apiEndpointHost, ".")
+
+	switch {
+	case len(domainLabels) < 2:
+		return domainLabels[0]
+	case len(domainLabels[0]) == 0 || len(domainLabels[1]) == 0:
+		return apiEndpointHost
+	default:
+		// Drop first label of domain (e.g. api.foobar.com -> foobar.com).
+		return strings.Join(domainLabels[1:], ".")
+	}
+}
+
+func findComponentVersion(component string, releaseComponents []releasev1alpha1.ReleaseSpecComponent) (string, error) {
```

Comment:
I think we have a function for this in the `key` package?

Reply:
I couldn't find. Did you possibly see it in `k8scloudconfig`: https://github.com/giantswarm/k8scloudconfig/blob/master/pkg/template/release.go#L66 ?

Reply:
I was thinking about https://github.com/giantswarm/azure-operator/blob/4471d37c30d2570f7ecf02d9701e0b37e0549cc2/service/controller/key/key.go#L191-L198

Reply:
Nice catch! Got it.